builder/simplify.cc

```diff
@@ -391,25 +425,44 @@ class simplifier : public node_mutator {
       set_result(let_stmt::make(std::move(lets), std::move(body)));
     }
   }
+  
+  void merge_loop_references(symbol_map<symbol_info> add) {
+    for (symbol_id i = 0; i < add.size(); ++i) {
+      if (!add[i]) continue;
+      std::optional<symbol_info>& info = references[i];
+      if (!info) {
+        info = std::move(add[i]);
+      } else {
+        info->ref_count += add[i]->ref_count;
+        info->referenced_in_loop = true;
```

Comment:
This should be `info->referenced_in_loop = info_referenced_in_loop || is_in_loop` (so if `is_in_loop` is false, a nested loop still marks references as in a loop).

Also, I think this needs to be outside this if (so it sets `referenced_in_loop` for symbols that didn't exist outside the loop).